src/audio/google/google_rtc_audio_processing.c

```diff
@@ -633,121 +633,139 @@ static int google_rtc_audio_processing_prepare(struct processing_module *mod,
 	return 0;
 }
 
+/* Extracts connected buffer pointers.  Clumsy, should cache in comp_data */
+static void find_bufs(struct processing_module *mod, struct comp_buffer **mic,
+		      struct comp_buffer **ref, struct comp_buffer **out)
+{
+	struct google_rtc_audio_processing_comp_data *cd = module_get_private_data(mod);
+	struct list_item *li;
+	int i = 0;
+
+	list_for_item(li, &mod->dev->bsource_list) {
+		struct comp_buffer *b = container_of(li, struct comp_buffer,
+						     sink_list);
+		if (i == cd->raw_microphone_source)
+			*mic = b;
+		else if (i == cd->aec_reference_source)
+			*ref = b;
+		i++;
+	}
+
+	*out = list_first_item(&mod->dev->bsink_list,
+			       struct comp_buffer, source_list);
+}
+
+static int trigger_handler(struct processing_module *mod, int cmd)
+{
+	struct comp_buffer *mic, *ref, *out;
+
+	find_bufs(mod, &mic, &ref, &out);
+
+	/* Ignore and halt propagation if we get a trigger from the
+	 * playback pipeline: not for us.
+	 */
+	if (ref->walking)
+		return PPL_STATUS_PATH_STOP;
+
+	/* Note: not module_adapter_set_state().  With IPC4 those are
+	 * identical, but IPC3 has some odd-looking logic that
+	 * validates that no sources are active when receiving a
+	 * PRE_START command, which obviously breaks for our reference
+	 * stream if playback was already running when our pipeline
+	 * started
+	 */
+	return comp_set_state(mod->dev, cmd);
+}
+
 static int google_rtc_audio_processing_reset(struct processing_module *mod)
 {
 	comp_dbg(mod->dev, "google_rtc_audio_processing_reset()");
-
 	return 0;
 }
 
+/* FunctionMostlyExistsToKeepLineLengthsUnderControl */
+static inline void execute_aec(struct google_rtc_audio_processing_comp_data *cd)
+{
+	/* FIXME: sample/frame format is platform dependent, these are
+	 * hard-configured format APIs and need indirection.  Note
+	 * that the calling code in process() is format-independent.
+	 */
+	GoogleRtcAudioProcessingAnalyzeRender_int16(cd->state,
+						    cd->aec_reference_buffer);
+	GoogleRtcAudioProcessingProcessCapture_int16(cd->state,
+						     cd->raw_mic_buffer,
+						     cd->output_buffer);
+	cd->raw_mic_buffer_frame_index = 0;
+}
+
 static int google_rtc_audio_processing_process(struct processing_module *mod,
 					       struct input_stream_buffer *input_buffers,
 					       int num_input_buffers,
 					       struct output_stream_buffer *output_buffers,
 					       int num_output_buffers)
```

Comment:
For those reviewing: note that this function was basically rewritten, but the diff is chunked up piecewise, making it harder to understand that it needs to be.  I'll move it in the next spin to keep them separate, but until then review in side-by-side mode for clarity.